### PR TITLE
Support pai and paiYarn trainingservice

### PR DESCRIPTION
**Pull request by SparkSnail**
test image `sparksnail/nni:paiYarn`

### CODE REVIEW

File: docs/en_US/TrainingService/PaiYarnMode.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
+**Run an Experiment on OpenpaiYarn**
+===
+The original `pai` mode is modificated to `paiYarnYarn` mode, which is a distributed training platform based on Yarn.
```
Comment (xuehui1991) [inline docs/en_US/TrainingService/PaiYarnMode.md]: "paiYarnYarn"?

File: docs/en_US/TrainingService/PaiMode.md [hunk 2]
```diff
@@ -53,52 +56,13 @@ Compared with [LocalMode](LocalMode.md) and [RemoteMachineMode](RemoteMachineMod
...
-    tensorboard --logdir . --port ${!TENSORBOARD_PORT} &
-    python3 mnist.py
-    ```
-    The config file of portList should be filled as following:
+* nniManagerNFSMountPath
+    * Required key. Set the mount path in your nniManager machine.
+* containerNFSMountPath
+    * Required key. Set the mount path in your container used in PAI.
+* paiStoragePlugin
+    * Required key. Set the sotrage plugin name used in PAI.
```
Comment (chicm-ms) [inline docs/en_US/TrainingService/PaiMode.md]: typo: sotrage

File: src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+import { PAIBaseJobInfoCollector } from './paiBaseJobInfoCollector';
+import { PAIBaseJobRestServer, ParameterFileMeta } from './paiBaseJobRestServer';
+import { PAIBaseClusterConfig, PAIBaseTrialJobDetail } from './paiBaseConfig';
+
+/**
+ * Training Service implementation for OpenPAI (Open Platform for AI)
+ * Refer https://github.com/Microsoft/pai for more info about OpenPAI
+ */
+@component.Singleton
+class PAIBaseTrainingService implements TrainingService {
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts]: Suggest to change this base class to abstract class.

File: src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+        this.logCollection = 'none';
+        this.log.info('Construct paiBase training service.');
+    }
+
+    public async run(): Promise<void> {
+        return;
+    }
+
+    public async submitTrialJob(form: TrialJobApplicationForm): Promise<any> {
+        return null;
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts]: Throw not implemented error.

File: src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+    }
+
+    public async updateTrialJob(trialJobId: string, form: TrialJobApplicationForm): Promise<TrialJobDetail> {
+        const trialJobDetail: undefined | TrialJobDetail = this.trialJobsMap.get(trialJobId);
+        if (trialJobDetail === undefined) {
+            throw new Error(`updateTrialJob failed: ${trialJobId} not found`);
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts]: This implementation sees incorrect.

File: src/nni_manager/training_service/pai_base/pai/paiTrainingService.ts [hunk 6]
```diff
@@ -0,0 +1,326 @@
+/**
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai_base/pai/paiTrainingService.ts]: why this `pai` dir is under `pai_base` dir?

File: src/nni_manager/training_service/pai_base/pai/paiTrainingService.ts [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+        super();
+    }
+
+    public async run(): Promise<void> {
+        this.log.info('Run PAI training service.');
+        const restServer: PAIJobRestServer = component.get(PAIJobRestServer);
+        await restServer.start();
+        restServer.setEnableVersionCheck = this.versionCheck;
+        this.log.info(`PAI Training service rest server listening on: ${restServer.endPoint}`);
+        await Promise.all([
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai_base/pai/paiTrainingService.ts]: This `run` is almost same as `PAIYarnTrainingService.run`,  should manage to move it up to base class

File: src/nni_manager/training_service/pai_base/pai/paiTrainingService.ts [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+
+        const deferred: Deferred<void> = new Deferred<void>();
+        const restServer: PAIJobRestServer = component.get(PAIJobRestServer);
+        try {
+            await restServer.stop();
+            deferred.resolve();
+            this.log.info('PAI Training service rest server stopped successfully.');
+        } catch (error) {
+            this.log.error(`PAI Training service rest server stopped failed, error: ${error.message}`);
+            deferred.reject(error);
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai_base/pai/paiTrainingService.ts]: Move this up to base class

File: src/nni_manager/training_service/pai_base/paiYarn/paiYarnTrainingService.ts [hunk 9]
```diff
@@ -471,9 +362,9 @@ class PAITrainingService implements TrainingService {
         return deferred.promise;
     }
 
-    private async statusCheckingLoop(): Promise<void> {
+    protected async statusCheckingLoop(): Promise<void> {
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai_base/paiYarn/paiYarnTrainingService.ts]: base class already has similiar one, should not be re-implemented like this.

File: docs/en_US/TrainingService/PaiYarnMode.md (hunk 1 above)
Comment (SparkSnail) [reply docs/en_US/TrainingService/PaiYarnMode.md]: fixed, changed to `paiYarn`

File: docs/en_US/TrainingService/PaiMode.md (hunk 2 above)
Comment (SparkSnail) [reply docs/en_US/TrainingService/PaiMode.md]: fixed.

File: src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts (hunk 3 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts (hunk 4 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai_base/pai/paiTrainingService.ts (hunk 8 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai_base/pai/paiTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai_base/paiYarn/paiYarnTrainingService.ts (hunk 9 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai_base/paiYarn/paiYarnTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai_base/pai/paiTrainingService.ts (hunk 6 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai_base/pai/paiTrainingService.ts]: `pai` and `paiYarn` are two trainingService, they inherit from `paiBase`

File: src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+import { PAIBaseJobInfoCollector } from './paiBaseJobInfoCollector';
+import { PAIBaseJobRestServer, ParameterFileMeta } from './paiBaseJobRestServer';
+import { PAIBaseClusterConfig, PAIBaseTrialJobDetail } from './paiBaseConfig';
+import { PAIJobRestServer } from './pai/paiJobRestServer';
+
+/**
+ * Training Service implementation for OpenPAI (Open Platform for AI)
+ * Refer https://github.com/Microsoft/pai for more info about OpenPAI
+ */
+@component.Singleton
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts]: Remove this `@component.Singleton`

File: src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts (hunk 5 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai_base/pai/paiTrainingService.ts (hunk 7 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai_base/pai/paiTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai/paiJobRestServer.ts [hunk 11]
```diff
@@ -21,17 +21,17 @@ export interface ParameterFileMeta {
  */
 @component.Singleton
 export class PAIJobRestServer extends ClusterJobRestServer {
-    private parameterFileMetaList: ParameterFileMeta[] = [];
+    protected parameterFileMetaList: ParameterFileMeta[] = [];
 
     @Inject
-    private readonly paiTrainingService: PAITrainingService;
+    protected readonly paiTrainingService: PAITrainingService;
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiJobRestServer.ts]: `@Inject` should be removed if it is  assigned in constructor.

File: src/nni_manager/training_service/pai/paiTrainingService.ts [hunk 12]
```diff
@@ -22,70 +22,96 @@ import { delay, generateParamFileName,
...
-    private paiRestServerPort?: number;
-    private nniManagerIpConfig?: NNIManagerIpConfig;
-    private copyExpCodeDirPromise?: Promise<void>;
-    private copyAuthFilePromise?: Promise<void>;
-    private versionCheck: boolean = true;
-    private logCollection: string;
-    private isMultiPhase: boolean = false;
-    private authFileHdfsPath: string | undefined = undefined;
-    private portList?: string | undefined;
+abstract class PAITrainingService implements TrainingService {
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiTrainingService.ts]: Remove this `@component.Singleton`

File: src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts (hunk 10 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai/paiJobRestServer.ts (hunk 11 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiJobRestServer.ts]: fixed.

File: src/nni_manager/training_service/pai/paiTrainingService.ts (hunk 12 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai/paiTrainingService.ts [hunk 13]
```diff
@@ -213,86 +192,6 @@ class PAITrainingService implements TrainingService {
...
-                break;
-            default:
-                //Reject for unknown keys
-                throw new Error(`Uknown key: ${key}`);
-        }
-
-        return deferred.promise;
-    }
-
     public getClusterMetadata(key: string): Promise<string> {
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiTrainingService.ts]: It seems this getClusterMetadata is not implemented, then throw not implemented error.

File: src/nni_manager/training_service/pai/paiTrainingService.ts [hunk 14]
```diff
@@ -304,11 +203,13 @@ class PAITrainingService implements TrainingService {
...
         this.stopping = true;
+        if (this.paiJobRestServer === undefined) {
+            throw new Error('paiBaseJobRestServer not initialized!');
+        }
 
         const deferred: Deferred<void> = new Deferred<void>();
-        const restServer: PAIJobRestServer = component.get(PAIJobRestServer);
         try {
-            await restServer.stop();
+            await this.paiJobRestServer.stop();
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiTrainingService.ts]: Remove this unnecessary `deferred`

File: src/nni_manager/training_service/pai/paiTrainingService.ts [hunk 15]
```diff
@@ -485,25 +238,11 @@ class PAITrainingService implements TrainingService {
                 }
             }
             await this.paiJobCollector.retrieveTrialStatus(this.paiToken, this.paiClusterConfig);
-            const restServer: PAIJobRestServer = component.get(PAIJobRestServer);
-            if (restServer.getErrorMessage !== undefined) {
-                throw new Error(restServer.getErrorMessage);
+            if (this.paiJobRestServer === undefined) {
+                throw new Error('paiBaseJobRestServer not implemented!');
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiTrainingService.ts]: error message is not accurate,  should be paiJobRestServer  is undefined.

File: src/nni_manager/training_service/pai/paiK8S/paiK8STrainingService.ts [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+@component.Singleton
+class PAIK8STrainingService extends PAITrainingService {
+    protected paiTrialConfig: NNIPAIK8STrialConfig | undefined;
+
+    constructor() {
+        super();
+    }
+
+    public async setClusterMetadata(key: string, value: string): Promise<void> {
+        const deferred: Deferred<void> = new Deferred<void>();
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiK8S/paiK8STrainingService.ts]: Remove the `deferred`

File: src/nni_manager/training_service/pai/paiK8S/paiK8STrainingService.ts [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
+                    break;
+                }
+                this.paiTrialConfig = <NNIPAIK8STrialConfig>JSON.parse(value);
+
+                // Validate to make sure codeDir doesn't have too many files
+                try {
+                    await validateCodeDir(this.paiTrialConfig.codeDir);
+                } catch (error) {
+                    this.log.error(error);
+                    deferred.reject(new Error(error));
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiK8S/paiK8STrainingService.ts]: this error should have already been logged by rest server,   no need to catch and re-throw, please check it.

File: src/nni_manager/training_service/pai/paiK8S/paiK8STrainingService.ts [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+            throw new Error(`updateTrialJob failed: ${trialJobId} not found`);
+        }
+        return trialJobDetail;
+    }
+
+    public async submitTrialJob(form: TrialJobApplicationForm): Promise<TrialJobDetail> {
+        if (this.paiClusterConfig === undefined) {
+            throw new Error(`paiClusterConfig not initialized!`);
+        }
+        const deferred: Deferred<PAITrialJobDetail> = new Deferred<PAITrialJobDetail>();
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiK8S/paiK8STrainingService.ts]: Remove this `deferred`

File: src/nni_manager/training_service/pai/paiTrainingService.ts [hunk 19]
```diff
@@ -101,13 +127,13 @@ class PAITrainingService implements TrainingService {
...
-        const paiTrialJob: PAITrialJobDetail | undefined = this.trialJobsMap.get(trialJobId);
+        const paiBaseTrialJob: PAITrialJobDetail | undefined = this.trialJobsMap.get(trialJobId);
 
-        if (paiTrialJob === undefined) {
+        if (paiBaseTrialJob === undefined) {
             return Promise.reject(`trial job ${trialJobId} not found`);
         }
 
-        return Promise.resolve(paiTrialJob);
+        return Promise.resolve(paiBaseTrialJob);
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiTrainingService.ts]: return `paiBaseTrialJob` directly

File: src/nni_manager/training_service/pai/paiTrainingService.ts [hunk 20]
```diff
@@ -101,13 +127,13 @@ class PAITrainingService implements TrainingService {
             throw new Error('PAI Cluster config is not initialized');
         }
 
-        const paiTrialJob: PAITrialJobDetail | undefined = this.trialJobsMap.get(trialJobId);
+        const paiBaseTrialJob: PAITrialJobDetail | undefined = this.trialJobsMap.get(trialJobId);
 
-        if (paiTrialJob === undefined) {
+        if (paiBaseTrialJob === undefined) {
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiTrainingService.ts]: throw new Error to keep style consistent

File: src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts [hunk 21]
```diff
@@ -0,0 +1,374 @@
...
+
+        // Step 2. Upload code files in codeDir onto HDFS
+        try {
+            await HDFSClientUtility.copyDirectoryToHdfs(trialLocalTempFolder, hdfsCodeDir, this.hdfsClient);
+        } catch (error) {
+            this.log.error(`PAI Training service: copy ${this.paiTrialConfig.codeDir} to HDFS ${hdfsCodeDir} failed, error is ${error}`);
+            trialJobDetail.status = 'FAILED'; // eslint-disable-line require-atomic-updates
+            deferred.resolve(true);
+
+            return deferred.promise;
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts]: return true directly

File: src/nni_manager/training_service/pai/paiTrainingService.ts (hunk 13 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai/paiTrainingService.ts (hunk 14 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai/paiTrainingService.ts (hunk 15 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai/paiK8S/paiK8STrainingService.ts (hunk 16 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiK8S/paiK8STrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai/paiK8S/paiK8STrainingService.ts (hunk 17 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiK8S/paiK8STrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai/paiK8S/paiK8STrainingService.ts (hunk 18 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiK8S/paiK8STrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai/paiTrainingService.ts (hunk 19 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai/paiTrainingService.ts (hunk 20 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts (hunk 21 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts]: fixed.

File: tools/nni_cmd/launcher_utils.py [hunk 22]
```diff
@@ -261,7 +262,7 @@ def validate_machine_list(experiment_config):
 
 def validate_pai_trial_conifg(experiment_config):
     '''validate the trial config in pai platform'''
-    if experiment_config.get('trainingServicePlatform') == 'pai':
+    if experiment_config.get('trainingServicePlatform') in ['pai', 'PAIYarn']:
```
Comment (chicm-ms) [inline tools/nni_cmd/launcher_utils.py]: It seems `PAIYarn` should be `paiYarn`

File: src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts [hunk 23]
```diff
@@ -0,0 +1,372 @@
...
+
+    constructor() {
+        super();
+    }
+
+    public async submitTrialJob(form: TrialJobApplicationForm): Promise<TrialJobDetail> {
+        if (this.paiClusterConfig === undefined) {
+            throw new Error(`paiBaseClusterConfig not initialized!`);
+        }
+        const deferred: Deferred<PAITrialJobDetail> = new Deferred<PAITrialJobDetail>();
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts]: Remove this deferred.

File: src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts [hunk 24]
```diff
@@ -0,0 +1,372 @@
...
+                    this.log.error('pai cluster config is not initialized');
+                    deferred.reject(new Error('pai cluster config is not initialized'));
+                    break;
+                }
+                this.paiTrialConfig = <NNIPAITrialConfig>JSON.parse(value);
+
+                // Validate to make sure codeDir doesn't have too many files
+                try {
+                    await validateCodeDir(this.paiTrialConfig.codeDir);
+                } catch (error) {
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts]: Remove this try catch

File: src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts [hunk 25]
```diff
@@ -0,0 +1,372 @@
...
+
+        this.trialJobsMap.set(trialJobId, trialJobDetail);
+        this.jobQueue.push(trialJobId);
+        deferred.resolve(trialJobDetail);
+
+        return deferred.promise;
+    }
+
+    public async setClusterMetadata(key: string, value: string): Promise<void> {
+        const deferred: Deferred<void> = new Deferred<void>();
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts]: Remove this deferred.

File: src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts [hunk 26]
```diff
@@ -0,0 +1,372 @@
...
+            experimentId: this.experimentId,
+            trialId: trialJobId,
+            filePath: hdfsHpFilePath
+        });
+    }
+
+    protected postParameterFileMeta(parameterFileMeta: ParameterFileMeta): Promise<void> {
+        const deferred: Deferred<void> = new Deferred<void>();
+        if (this.paiJobRestServer === undefined) {
+            throw new Error('paiBaseJobRestServer not implemented!');
```
Comment (chicm-ms) [inline src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts]: Error messge inaccurate.

File: tools/nni_cmd/launcher_utils.py (hunk 22 above)
Comment (SparkSnail) [reply tools/nni_cmd/launcher_utils.py]: fixed.

File: src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts (hunk 23 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts (hunk 24 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts (hunk 25 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts]: fixed.

File: src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts (hunk 26 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/pai/paiYarn/paiYarnTrainingService.ts]: fixed.

**Pull request merged**